### ISSUE TITLE
Difference between real Training Time Consumption and paper reported

### ISSUE THREAD

**Issue by yueyang130** [mila-iqia/spr#28]
First, I note that in the code backward propogation is done on CPU (see [it](https://github.com/mila-iqia/spr/blob/0b9dd4e7b9bbdfaecdf9a3713bf5931fb54ab0ca/src/algos.py#L302) ), which really takes more #CPUs and slow down the training. In this way, it takes me 1 A100 GPU and 8 CPUs to train about 30 hours.

After I move backward propogation to GPU, it still takes me 1 A100 GPU and 8 CPUs to train about 10 hours. However, in the paper, it says "SPR with augmentation for a 100K steps on Atari takes around 4 and a half to finish a complete training and evaluation run on a single game. We find that using data augmentation adds an overhead, and SPR without augmentation can run in just 3 hours." And you said in other issue (see [it](https://github.com/mila-iqia/spr/issues/23#issuecomment-794794990) ), it needs a V100 to train 6 hours at very most.

So here is my questions:
1. Why did you do backward propogation on CPU in the code? 
2. The results you report( it needs a V100 to train 6 hours at very most) is doing backward on CPU or GPU?
3. Why there is difference between time you report and I actually take?
4.  How many CPUs should I take?
5. Can I get training done in shorter time by using more than one A100 even though I found one A 100 is not fully utilized when training?

I am a freshman on RL and above questions confuse me a lot. Any reply will be appreciated very much!!!

**Comment by yueyang130** [comment]
Additionally, how can I set the script to train with multiple GPUs?

**Comment by MaxASchwarzer** [comment]
I think you're confusing our code with EfficientZero; SPR is the base SSL method that EfficientZero uses, but our code implements DQN rather than MuZero as its RL algorithm. Our code should take at most a few hours on a single GPU, not seven on four GPUs; that number comes from EfficientZero. We have not added support for multi-GPU training, as we found it unnecessary when single-GPU training is so quick. I imagine that using multiple GPUs would reduce training time, but not by a factor of four, since our training time also includes environment interaction and replay sampling.

As for putting the losses on the CPU, that structure is inherited from rlpyt and doesn't actually do backprop on CPU, at least in the version of Torch we used. As far as we're aware all meaningful computations were done on GPU.

Honestly, I have no idea why your code is taking so much longer to run than ours. We're planning on releasing a Jax version of SPR in the very near future; perhaps that will run better on your system.
